File: leetcode-75/1071_greatest_common_divisor_of_string/greatest_common_divisor_of_string.py
class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        # check if the strings are equal in both order, if not return empty string
        if str1 + str2 != str2 + str1:
            return ""

        # return the substring of GCD length equal to the two string 
        return str1[:self.gcd(len(str1), len(str2))]

    def gcd(self, a: int, b: int) -> int:
        while b:
            a, b = b, a % b
        return a


solution = Solution()
print(solution.gcdOfStrings("ABCABC", "ABC"))  # Output: "ABC"
print(solution.gcdOfStrings("ABABAB", "ABAB"))  # Output: "AB"
print(solution.gcdOfStrings("LEET", "CODE"))    # Output: ""

File: leetcode-75/1071_greatest_common_divisor_of_string/greatest_common_divisor_of_string.rs
struct Solution;

impl Solution {
    pub fn gcd_of_strings(str1: String, str2: String) -> String {
        // Check if the concatenated strings are equal in both orders
        if format!("{}{}", &str1, &str2) != format!("{}{}", &str2, &str1) {
            return String::new(); 
        }

        // Compute the GCD of the lengths of the two strings
        let gcd_length = gcd(str1.len(), str2.len());

        // Return the substring of length equal to the GCD
        str1[..gcd_length].to_string()
    }
}

// Helper function to compute the GCD of two numbers using the Euclidean algorithm
fn gcd(a: usize, b: usize) -> usize {
    let mut a = a;
    let mut b = b;
    while b != 0 {
        let temp = b;
        b = a % b;
        a = temp;
    }
    a
}

fn main() {
    println!("{}", Solution::gcd_of_strings("ABCABC".to_string(), "ABC".to_string())); // Output: "ABC"
    println!("{}", Solution::gcd_of_strings("ABABAB".to_string(), "ABAB".to_string())); // Output: "AB"
    println!("{}", Solution::gcd_of_strings("LEET".to_string(), "CODE".to_string()));   // Output: ""
}
